### PR TITLE
Migrate memory to ChatConversationHistory()

### LINKED ISSUE #44: Migrate to `ChatMessageHistory()`
Ayush pointed out that serializing the history into one string doesn't utilize the chat API that OpenAI has developed. Langchain added support for `ChatMessageHistory()` that organizes previous messages much better that we can migrate to now.

### PR DESCRIPTION
**Pull request by jacobvm04**
Fixes #44 by switching from ConversationBufferMemory to ChatConversationHistory. 

For code readability, `bloom_chain.py` is created to house the logic for generating thoughts and responses, and is used in `chain.py`.

### CODE REVIEW

File: agent/data/prompts/objective/system/response.yaml [hunk 1]
```diff
@@ -1,9 +1,8 @@
...
+  You are Bloom, a subversive-minded learning companion with deep expertise in Socratic dialogue. Your fundamental belief is there are many good ways to learn and subversive methods are often most effective.  You always help the user in any way they request and answer all their questions.
 
-    {thought}
-    
-    You must produce an appropriate response to the user input. Keep your responses concise and specific, always end each response with ONLY ONE topically relevant questions that drives the conversation forward, and if the user wants to end the conversation, always comply.
+  You must produce an appropriate response to the user input. Keep your responses concise and specific, always end each response with ONLY ONE topically relevant questions that drives the conversation forward, and if the user wants to end the conversation, always comply.
+
+  It is important that you pay a lot of attention to the "thought" in the conversation!
```
Comment (vintrocode) [inline agent/data/prompts/objective/system/response.yaml:8]: was it getting lost in the context window?

**Comment by vintrocode** [comment]
[96710f5](https://github.com/plastic-labs/tutor-gpt/pull/45/commits/96710f5c239b355f2f7d99812cffeb78533eee13) closes  #40 which was left over from when we were trying to discern conversation types (spoiler: [theory of mind is all you need](https://plasticlabs.ai/blog/Theory-of-Mind-is-All-You-Need))

File: agent/data/prompts/objective/system/response.yaml (hunk 1 above)
Comment (jacobvm04) [reply agent/data/prompts/objective/system/response.yaml]: Hm, I think was from so earlier debugging I was doing, but imo it shouldn't be necessary now, I'll make a commit to remove it

File: bot/core.py [hunk 2]
```diff
@@ -19,23 +18,21 @@ def __init__(self, bot) -> None:
...
+            thought_chain=bloom_chain,
             thought_memory=local_chain.thought_memory
         )
         response = await chat(
             inp=input,
             thought=thought,
-            response_chain=response_chain,
+            response_chain=bloom_chain,
             response_memory=local_chain.response_memory
         )
```
Comment (VVoruganti) [inline bot/core.py:34]: So if we changed the interface in bloom chain to have `think` and `respond` methods is there a reason to have two different calls to chat. I would recommend we can refactor chat to automatically return the thought and response under the hood. 

I'm guessing we did this for backwards compatibility sake, but at this stage not a huge concern. Rather the code be clean.

File: agent/bloom_chain.py
Comment (VVoruganti) [inline agent/bloom_chain.py:1]: I don't really see the reason to split this into it's own file.

File: agent/chain.py [hunk 3]
```diff
@@ -64,52 +29,35 @@ def load_chains():
...
-
-    objective_thought_chat_prompt = ChatPromptTemplate.from_messages([objective_system_thought, objective_human_thought])
-    objective_response_chat_prompt = ChatPromptTemplate.from_messages([objective_system_response, objective_human_response])
-
-    # define chains
-    objective_thought_chain = LLMChain(
-        llm=llm,
-        prompt=objective_thought_chat_prompt,
+    # define chain
+    bloom_chain = BloomChain(
```
Comment (VVoruganti) [inline agent/chain.py:33]: This is essentially a constructor for `BloomChain` with some pre-determined hardcoded inputs. I think generally the `agent/` folder should more be general interfaces and classes with concrete instances in the `bot/` and `www/` folders. We could make this a overloaded constructor within the class itself as a shortcut constructor.

TL;DR  - I think this should stay within the class instead of separate

File: agent/chain.py [hunk 4]
```diff
@@ -64,52 +29,35 @@ def load_chains():
...
-    )
-
-    return ( 
-        objective_thought_chain, 
-        objective_response_chain,
-    )
+    return bloom_chain
 
 
 async def chat(**kwargs):
```
Comment (VVoruganti) [inline agent/chain.py:41]: See below comment on changes to `chat`. TL;DR is that we can make this another method within the `BloomChain` class that automatically does the thought response loop instead of requiring two separate calls.

File: agent/chain.py [hunk 5]
```diff
@@ -2,60 +2,25 @@
...
-# OBJECTIVE_SYSTEM_THOUGHT = load_prompt("./data/prompts/objective/system/thought.yaml")
-# OBJECTIVE_SYSTEM_RESPONSE = load_prompt("./data/prompts/objective/system/response.yaml")
-# OBJECTIVE_HUMAN_THOUGHT = load_prompt("./data/prompts/objective/human/thought.yaml")
-# OBJECTIVE_HUMAN_RESPONSE = load_prompt("./data/prompts/objective/human/response.yaml")
-# OBJECTIVE_SUMMARY_THOUGHT = load_prompt("data/prompts/objective/summaries/thought.yaml")
-# OBJECTIVE_SUMMARY_RESPONSE = load_prompt("data/prompts/objective/summaries/response.yaml")
 
 
-def load_memories(conversation_type: str = "objective"):
+def load_memories():
```
Comment (VVoruganti) [inline agent/chain.py:15]: While we are making this more object oriented with classes we may as well move this into the `ConversationCache` class since it's really only used within the constructor. 

Can make it a class method as well if we want to expose the interface, but don't see a big reason right now.

File: www/main.py [hunk 6]
```diff
@@ -79,23 +75,21 @@ def init():
 thought, response = '', ''
 async def chat_and_save(local_chain: ConversationCache, input: str) -> tuple[str, str]:
         global thought, response
-        thought_chain =  OBJECTIVE_THOUGHT_CHAIN 
-        response_chain = OBJECTIVE_RESPONSE_CHAIN # if local_chain.conversation_type == "discuss" else WORKSHOP_RESPONSE_CHAIN
+        bloom_chain =  BLOOM_CHAIN # if local_chain.conversation_type == "discuss" else WORKSHOP_RESPONSE_CHAIN
         # response_chain = local_chain.conversation_type == "discuss" ? DISCUSS_RESPONSE_CHAIN : WORKSHOP_RESPONSE_CHAIN
 
         thought = await chat(
```
Comment (VVoruganti) [inline www/main.py:81]: Same as above

**Review by VVoruganti — CHANGES_REQUESTED** [review]
Overall I like the direction you are going just think there's a bit more room to simplify further. Can keep all of the agent stuff in one file still don't think we need `chain` and `bloom_chain` as separate files. 

Move to keep everything we can within the classes and not have un-parented methods.

**Review by VVoruganti — APPROVED** [review]
Put everything into one file, made `think` and `respond` async methods again, fixed bug in `restart` method in `bot/core.py`. Tested on my local server and looks good
